[exporter/elasticsearch] Flesh out encoding log records with `mapping.mode: ecs`

ycombinator opened the pull request:
**Description:** <Describe what has changed.>

This PR is a follow up to https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/31553.  In that PR, the foundations were laid for encoding log records with ECS fields when the user specifies `mapping.mode: ecs` in their exporter configuration.

This PR adds ECS conversions for several additional fields.

I expect there will be more followup PR(s) to add more conversions, especially ones that might not be straightforward.

**Link to tracking issue:** #29742 

**Testing:** Added unit tests

**Documentation:** The `mapping.mode: ecs` configuration setting is already [documented](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/exporter/elasticsearchexporter#configuration-options).

ycombinator committed:
Encode logs based on mapping mode

ycombinator committed:
WIP: encode log in ECS mode

ycombinator committed:
Adding CHANGELOG entry

ycombinator committed:
Adding TODO

ycombinator committed:
Fleshing out some more

ycombinator committed:
Add TODO

ycombinator committed:
Add stub for unit test

ycombinator committed:
Use consts from semconv package for attribute names

ycombinator committed:
Pass now for testability

ycombinator committed:
Flesh out unit test

axw commented on exporter/elasticsearchexporter/model.go:
```diff
@@ -137,3 +182,17 @@ func scopeToAttributes(scope pcommon.InstrumentationScope) pcommon.Map {
 }
+
+func mapLogAttributesToECS(document *objmodel.Document, attrs pcommon.Map, conversionMap map[string]string) {
+	attrs.Range(func(k string, v pcommon.Value) bool {
```
Internally `pcommon.Map` is an array of key-value pairs (https://github.com/open-telemetry/opentelemetry-collector/blob/f0473ca0a55293b92374cc77e0275026537342da/pdata/pcommon/map.go#L26), so even if `conversionMap` is smaller we'd likely end up iterating `attrs` multiple times that way.

ycombinator replied:
I did add an early return for the special case where `conversionMap` was empty: 47905d4e98ddccb243dcc3eadd2ac8ed1dc6e097.

ycombinator committed:
Remove passthrough keys from conversion map

ycombinator committed:
If conversion map is empty, just add all attributes

ycombinator committed:
Add unit test for mapLogAttributesToECS

ycombinator commented on exporter/elasticsearchexporter/model.go:
```diff
@@ -58,9 +73,39 @@ func (m *encodeModel) encodeLog(resource pcommon.Resource, record plog.LogRecord
+	// First, try to map resource-level attributes to ECS fields.
+	resourceAttrsConversionMap := map[string]string{
+		semconv.AttributeServiceName:           "service.name",
+		semconv.AttributeServiceVersion:        "service.version",
```
++ done in 3ea4512c3e3451d73708fbed7d0ff2576cf3b304.

ycombinator committed:
Quote changelog string with backticks

ycombinator committed:
Flesh out more conversions

ycombinator committed:
Correctly populate @timestamp

ycombinator committed:
Rename helper function

ycombinator committed:
Extract helper function

ycombinator committed:
Handle event.name -> event.action

ycombinator committed:
Fixing linter error

axw commented on exporter/elasticsearchexporter/model.go:
```diff
@@ -58,9 +73,47 @@ func (m *encodeModel) encodeLog(resource pcommon.Resource, record plog.LogRecord
+		semconv.AttributeHostName:              "host.hostname",
+		semconv.AttributeHostArch:              "host.architecture",
+		semconv.AttributeProcessExecutablePath: "process.executable",
+		semconv.AttributeOSType:                "os.platform",
```
Should we be vetting the values here? ECS expects `os.platform` to have one of a restricted set of values -- https://www.elastic.co/guide/en/ecs/current/ecs-os.html#field-os-type

For iOS and Android, I think we should set `os.platform` from `os.name` ("Android" or "iOS") rather than `os.type` ("linux" or "darwin")

See https://github.com/elastic/apm-data/blob/0e072ceaadae2ef8f61d75c3a27c47cc5ce81257/input/otlp/metadata.go#L320-L338

ycombinator replied:
Implemented in d36a76fff8.

axw reviewed (commented):
Mostly LGTM, just a few comments so far.

felixbarny commented on exporter/elasticsearchexporter/model.go:
```diff
@@ -58,9 +73,47 @@ func (m *encodeModel) encodeLog(resource pcommon.Resource, record plog.LogRecord
+	var document objmodel.Document
+
+	// First, try to map resource-level attributes to ECS fields.
+	resourceAttrsConversionMap := map[string]string{
```
I think what's missing is the translation of `k8s.*` SemConv resource attributes. See also https://github.com/elastic/apm-data/blob/0e072ceaadae2ef8f61d75c3a27c47cc5ce81257/input/otlp/metadata.go#L145-L165

ycombinator replied:
Yep, I didn't include them as they're not really part of ECS.  But as per the off-PR discussion with @axw and you, I will include them now for practical reasons.

ycombinator replied:
Added in 468043f049.

ycombinator replied:
Done in 468043f0494859176c0fd89d3c1d6c62466068d2.

carsonip commented on exporter/elasticsearchexporter/model.go:
```diff
@@ -40,6 +41,20 @@ const (
+	var document objmodel.Document
+	switch m.mode {
+	case MappingECS:
+		document = m.encodeLogECSMode(resource, record, scope, time.Now())
```
I imagine the closest one to event.received is observedTimestamp: https://opentelemetry.io/docs/specs/otel/logs/data-model/#field-observedtimestamp

I expect the receiver to set observedTimestamp, and the exporter to use it.

lahsivjar replied:
> Perhaps we should omit the explicit mapping of event.received, and leave it as a dynamically mapped attribute? i.e. leave it to the receiver or some upstream client to set the event.received attribute.

+1 on this. The observed timestamp should be a fallback for the `@timestamp` field if the record doesn't have the timestamp (the PR already does that in `encodeLogTimestampECSMode` method).

felixbarny replied:
There's a bit of a difference when this logic gets executed. In APM Server, this is executed when receiving OTLP events. However, when doing it in the ES exporter it's set when exporting/sending events to ES. The difference is that there may be a delay in between receiving an exporting the events, especially if the events sit in a queue for a while. So to do it in a similar way as APM Server, we'd need to set `event.received` in the OTLP receiver. But that's not in the scope of improving the ES exporter.

ycombinator replied:
Thanks for the clarification, @felixbarny.  Makes sense why the same logic does not apply in both scenarios.  Removed `event.received` setting code in d9e4f69a6e.

ycombinator committed:
Do not set event.received

ycombinator committed:
Add k8s.* -> kubernetes.* conversions

ycombinator committed:
Adding process.runtime.* -> service.runtime.* conversions

ycombinator committed:
Extract out resourceAttrsConversionMap to package-level variable

ycombinator commented on exporter/elasticsearchexporter/model.go:
```diff
@@ -58,9 +73,47 @@ func (m *encodeModel) encodeLog(resource pcommon.Resource, record plog.LogRecord
+		semconv.AttributeProcessExecutablePath: "process.executable",
+		semconv.AttributeOSType:                "os.platform",
+		semconv.AttributeOSDescription:         "os.full",
+	}
```
I moved this mapping out of this function and into package scope in bd90fbf4076586501b57b7799e6b12797cd30828 so, in the future, conversion code for other signals can use it too.

ycombinator committed:
No need to pass time.Now() anymore

ycombinator committed:
Implement ability to skip keys during conversion

ycombinator committed:
Handle agent name and service language name

ycombinator commented on exporter/elasticsearchexporter/model.go:
```diff
@@ -58,9 +73,47 @@ func (m *encodeModel) encodeLog(resource pcommon.Resource, record plog.LogRecord
+	resourceAttrsConversionMap := map[string]string{
+		semconv.AttributeServiceInstanceID:     "service.node.name",
+		semconv.AttributeDeploymentEnvironment: "service.environment",
+		semconv.AttributeTelemetrySDKName:      "agent.name",
```
Implemented this logic, including that for `telemetry.distro.name`, in 4b9cca0bdf.  Would you mind taking a look at the unit test cases, particularly the ones in `TestEncodeLogECSModeAgentName` to make sure I got the logic right?  Thanks!

ycombinator committed:
Handle conversion to ECS os.type field

ycombinator committed:
Fixing OS conversions

felixbarny commented on exporter/elasticsearchexporter/model.go:
```diff
@@ -137,3 +208,109 @@ func scopeToAttributes(scope pcommon.InstrumentationScope) pcommon.Map {
+	// Set agent name in document.
+	document.AddString("agent.name", agentName)
+
+	// Set service language name in document.
+	if telemetrySdkLanguage == "" {
+		telemetrySdkLanguage = "unknown"
+	}
+	document.AddString("service.language.name", telemetrySdkLanguage)
```
> One thing I think we should not carry across from apm-data is setting service.language.name when telemetry.sdk.language is unspecified. 

Sounds good. But maybe we have to adapt some of the assumptions we have in the APM UI about this field always being present.

> Maybe we can just skip setting service.language.name altogether for now, since it's included in agent.name?

The `service.language.name` attribute is used a lot in our UI. Since it directly translates to `service.language.name` I think I'd just add it to the `resourceAttrsConversionMap`.

axw replied:
> The service.language.name attribute is used a lot in our UI. Since it directly translates to service.language.name I think I'd just add it to the resourceAttrsConversionMap.

This PR is focused on logs though, and I think the `service.language.name` usage in the (APM) UI is not relevant here. Could we start without it, and add it in later if there's a need?

felixbarny replied:
let's do that

ycombinator replied:
Removed code for setting `service.language.name` in cb5a3b88bb.

ycombinator committed:
Don't set service.language.name

ycombinator committed:
Handle agent version

ycombinator commented on exporter/elasticsearchexporter/model.go:
```diff
@@ -6,16 +6,48 @@ package elasticsearchexporter // import "github.com/open-telemetry/opentelemetry
+	semconv.AttributeTelemetrySDKName:      "",
+	semconv.AttributeTelemetryDistroName:   "",
+	semconv.AttributeTelemetrySDKLanguage:  "",
+	semconv.AttributeTelemetrySDKVersion:   "agent.version",
```
Implemented in 04a681bd49.

ycombinator commented on exporter/elasticsearchexporter/model.go:
```diff
@@ -6,16 +6,48 @@ package elasticsearchexporter // import "github.com/open-telemetry/opentelemetry
+	semconv.AttributeServiceInstanceID:     "service.node.name",
+	semconv.AttributeDeploymentEnvironment: "service.environment",
+	semconv.AttributeTelemetrySDKName:      "",
+	semconv.AttributeTelemetryDistroName:   "",
```
Implemented as part of 04a681bd49.

axw reviewed (approved):
LGTM!

JaredTan95 commented:
wait https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/31553 get merged.

carsonip commented on .chloggen/exporter-elasticsarch-more-ecs-mode.yaml:
```diff
@@ -0,0 +1,27 @@
+note: "Converts more SemConv fields in OTel events to ECS fields in Elasticsearch documents when `mapping.mode: ecs` is specified."
+
+# Mandatory: One or more tracking issues related to the change. You can use the PR number here if no issue exists.
+issues: [29742]
```
```suggestion
issues: [31694]
```

ycombinator committed:
Mention SemConv version in README

ycombinator committed:
Update test

ycombinator committed:
Incorporate changes from prereq PR

ycombinator commented:
Hi @JaredTan95, I've rebased this PR on `main` and addressed conflicts after it's pre-requisite PR, #31553, was merged yesterday.  I've also incorporated the changes made in #31553 into the structure being used by the code in this PR (see 3533ef0cb094afbee112cb104178fb7659268da5).  

Would you mind re-reviewing this PR when you get a chance?  Thanks!

JaredTan95 reviewed (approved):
@ycombinator Thank you very much for your change and letting me know, I am glad to review, and changes LGTM.

ycombinator commented:
@JaredTan95 The required CI check, https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/8641453852/job/23691083097?pr=31694, is failing on this PR but the failure seems unrelated to the changes in this PR:

> Failed while attempting to install arkade. Please manually install:
> ...

What is normally done in such cases?  Is there some way to re-run CI on this PR?

ycombinator commented:
CI is [failing](https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/8780508502/job/24090919421?pr=31694#step:8:151) on an unrelated unit test that's also [failing](https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/8780289315/job/24090007607#step:8:152) on `main`.  @open-telemetry/collector-contrib-maintainer, could one of you please review this PR and merge if you think it's okay.  The codeowner, @JaredTan95, has already approved the PR.  Thanks!

JaredTan95 commented:
> CI is [failing](https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/8780508502/job/24090919421?pr=31694#step:8:151) on an unrelated unit test that's also [failing](https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/8780289315/job/24090007607#step:8:152) on `main`. @open-telemetry/collector-contrib-maintainer, could one of you please review this PR and merge if you think it's okay. The codeowner, @JaredTan95, has already approved the PR. Thanks!
> 
> [EDIT] Here's the issue tracking the unrelated failing test: #32391

Thanks for continuing to follow up, but we should wait until the main branch fixes this to make sure all CI passes.

andrzej-stencel committed:
update issue reference in changelog entry

BenEfrati commented:
Hi @ycombinator , @andrzej-stencel 

There is a regression after this change - duplicate key in json:
before this change request to ELK was:
```
{
...
   "host":{
      "name":"hostname"
   },
   "os":{
      "description":"Windows 11 10.0",
      "type":"windows"
   },
...
}
```

after this change:
```
{
...
   "host":{
      "hostname":"hostname",
      "os":{
         "full":"Windows 11 10.0",
         "platform":"windows"
      }
   },
...
 "host":{
      "os":{
         "type":"windows"
      }
   },
...
}
```

in this example `host` key is duplicated, result in invalid json and error from ELK
```
 "error": {
  "reason": "failed to parse",
  "type": "mapper_parsing_exception"
  },
  "status": 400
```

Using same resource attributes:
```
Resource attributes:
host.name: Str(hostname)
os.description: Str(Windows 11 10.0)
os.type: Str(windows)
```
_this is from debug exporter_

BenEfrati commented on exporter/elasticsearchexporter/model.go:
```diff
@@ -193,3 +230,117 @@ func scopeToAttributes(scope pcommon.InstrumentationScope) pcommon.Map {
+	if ecsHostOsType == "" {
+		return
+	}
+	document.AddString("host.os.type", ecsHostOsType)
```
in case host is already part of document it'll result in duplicate host key in json request:
Resource attributes:
```
host.name: Str(hostname)
os.description: Str(Windows 11 10.0)
os.type: Str(windows)
```
